core/src/main/java/org/apache/iceberg/hadoop/ConfigProperties.java

```diff
@@ -25,4 +25,6 @@ private ConfigProperties() {
   }
 
   public static final String ENGINE_HIVE_ENABLED = "iceberg.engine.hive.enabled";
+
+  public static final String CUSTOM_FILE_IO_IMPL = "iceberg.fileio.impl";
```

Comment:
Why do we need a Hadoop config property?

Reply:
I see this is used to get the implementation class. Iceberg should never use a Hadoop `Configuration` for config, except when integrating with an engine that uses it for config. It's okay to store configuration for Hive, but not for Iceberg core.

For `FileIO` implementation, I think the config should come from the catalog property map.

Reply:
Thank you for the comments! I will fix the other parts accordingly. I think this is the place I do not fully understand. Is there a catalog property map? I don't see such a thing in the catalog interface, and that is why I have to use the Hadoop config.

Reply:
There are config properties from both Flink and Spark that we use when constructing the Hive and Hadoop catalogs. And for #1640, we are talking about using an initialization method to set the config using a string map (and a catalog name). While these classes don't pull their own config out of a map, I think of the config as coming from a map of config properties.

Sorry for the confusion. I think we should configure this however the catalog is configured, which will very likely be a string map passed to an init method in the future.